Puzzle
----

https://thefiddler.substack.com/p/can-you-sprint-to-the-finish

**Fiddler**

This time around, you and a competitor are approaching the finish of a grueling stage of the Tour de Fiddler. One of you will win the stage, the other will come in second. As you approach the finish, each of you will test the feeling of your legs, which will be somewhere between 0 percent (“I can barely go on!”) and 100 percent (“I can do this all day!”). For the purposes of this puzzle, these values are chosen randomly, uniformly, and independently.

Immediately after feeling your legs, you and your opponent each have a decision to make. Do you maintain your current pace, or do you sprint to the finish? Among those who sprint for the finish, whoever’s legs are feeling the best will win the stage. But if no one sprints for the finish, everyone has an equal chance of winning the stage. In the Tour de Fiddler, you must each decide independently whether to sprint for the finish based on your legs—you don’t have time to react to your opponent’s decision.

Normally, teams at the Tour de Fiddler keep their strategy and tactics close to the vest. But earlier today, your opponent’s manager declared on international television that if (and only if) your opponent’s legs were feeling 50 percent or better, they’d sprint for the finish.

As you are about to test your legs for the final sprint and see how they feel, what are your chances of winning the stage, assuming an optimal strategy?

**Extra Credit**

Instead of one opponent, now you have two—meaning three riders in total. As luck would have it, the managers for both other riders proclaimed that they’d sprint for the finish if (and only if) their legs were feeling 50 percent or better. Note that your opponents’ feelings are independent of each other.

As the three of you near the finish, your own team manager radios you the following message: “If your legs feel 'garbled' percent or better, sprint for the finish!”

You can’t make out what the garbled part of the message is, and you’re too tired to radio back for confirmation. Instead, you somehow muster the energy to randomly, uniformly pick a number between 0 and 100 to fill in the blank from your manager’s message, thereby determining your racing strategy—optimization be damned!

Right before you choose your random strategy and test your legs, what are your chances of winning the stage against both opponents?

Fiddler Solution
----

Let the riders be $A$ (you) and $B$. Let the corresponding leg feels be $a$ and $b$.

First, we have to determine the optimal strategy.

Your leg feeling is $a$. You don't know anything about your opponent's leg feel, so it could be anywhere in the range $0$ to $1$.

If $a < 1/2$:
* If you don't sprint:
  * your chances of winning are $1/4$ (for you to win, your opponent has to not sprint (prob $1/2$), and you have to win the non sprint race (prob $1/2$))

* If you do sprint:
  * your chances of winning are $1/2$ (your opponent has to not sprint (because if they do sprint, their leg feel is definitely superior))

If $a >= 1/2$:
* If you don't sprint:
  * your chances of winning are $1/4$ (for you to win, your opponent has to not sprint (prob $1/2$), and you have to win the non sprint race (prob $1/2$))

* If you do sprint:
  * your chances of winning are $a$ (either your opponent does not sprint(prob $1/2$) or they do sprint, but their leg feel is worse(prob $a-1/2$)). Note that $a >= 1/2$ and that the average value of $a$ in this case is $3/4$ (average from 1/2 to 1).

So, in both cases your best option is to sprint.

Given that, your total probability of winning is $1/2 \times 1/2 + 1/2 \times 3/4$.

This is equal to $5/8$ = **0.625**


Extra Credit Solution
----

Let the riders be $A$ (you) ,  $B$ and $C$. Let the corresponding leg feels be $a$ , $b$ and $c$.

Let the threshold chosen randomly by you be $t$. And for convenience, $t' = 1-t$

Here, there's no strategy, but there are 8 cases, corresponding to $A,B,C$ sprinting or not.

* No sprinters
  * $P_{case} = t \times 1/2 \times 1/2 = t/4$
  * $P_{A_{win}} = 1/3 $

* Only B sprints 
  * $P_{case} = t \times 1/2 \times 1/2 = t/4$
  * $P_{A_{win}} = 0 $

* Only C sprints 
  * $P_{case} = t \times 1/2 \times 1/2 = t/4$
  * $P_{A_{win}} = 0 $

* B and C sprint
  * $P_{case} = t \times 1/2 \times 1/2 = t/4$
  * $P_{A_{win}} = 0$

* Only A sprints 
  * $P_{case} = (1-t) \times 1/2 \times 1/2 = t'/4$
  * $P_{A_{win}} = 1 $

* A and B sprint
  * $P_{case} = (1-t) \times 1/2 \times 1/2 = t'/4$
  * If $t < 1/2$
    * $P_{A_{win}} = (1/8)/(t'/2) = 1/4t' $
  * If $t >= 1/2$
    * $P_{A_{win}} = 1 - ((t'^2/2)/(t'/2)) = 1 - t' = t $

* A and C sprint
  * $P_{case} = (1-t) \times 1/2 \times 1/2 = t'/4$
  * If $t < 1/2$
    * $P_{A_{win}} = (1/8)/(t'/2) = 1/4t' $
  * If $t >= 1/2$
    * $P_{A_{win}} = 1 - ((t'^2/2)/(t'/2)) = 1 - t' = t $

* A, B and C sprint
  * $P_{case} = (1-t) \times 1/2 \times 1/2 = t'/4$
  * If $t < 1/2$
    * $P_{A_{win}} = (1/3)\times((1/8)/(t'/4)) = 1/6t' $
  * If $t >= 1/2$
    * $P_{A_{win}} = [1 \times (1/2-t')^2 \times t' + 2 \times 1/2 \times (1/2-t')t'^2  + 1/3 \times t'^3] / (t'/4)$
    * $P_{A_{win}} = (4/t') [ (1/2-t')^2 \times t' + (1/2-t')t'^2  + 1/3 \times t'^3 ] $
    * $P_{A_{win}} = 4 [ (1/2-t')^2  + (1/2-t')t' + 1/3 \times t'^2 ]$
    * $P_{A_{win}} = 4 [ (1/2-t')\times(1/2) + 1/3 \times t'^2 ]$
    * $P_{A_{win}} = 4 [ 1/4 - t'/2  + t'^2/3 ]$
    * $P_{A_{win}} = 4 [ 1/4 - (1-t)/2 + (1-t)^2/3 ]$
    * $P_{A_{win}} = 4 [ 1/4 - 1/2 + t/2 + 1/3 - 2t/3 + t^2/3 ]$
    * $P_{A_{win}} = 4 [ 1/12 - t/6 + t^2/3 ]$
    * $P_{A_{win}} = 1/3 - 2t/3 + 4t^2/3 $
    * $P_{A_{win}} = (1-2t+4t^2)/3 $
      * -- seems correct because it is 1/3 for t=0.5. 1 for t = 1 also seems okay (a will be 1 when this case happens).


Putting it all together, $P_{A_{win}}$ is:
* If $t < 1/2$
  * $P_{A_{win}} = (t/4) \times (1/3) + (t'/4) \times [1 + 1/2t' + 1/6t'] $
  * $P_{A_{win}} = (t/4) \times (1/3) + (t'/4) \times [(6t' + 3 + 1)/6t'] $
  * $P_{A_{win}} = (t/4) \times (1/3) + (t'/4) \times [(6t' + 4)/6t'] $
  * $P_{A_{win}} = (t/12) + (3t' + 2)/12 $
  * $P_{A_{win}} = (t + 3t' + 2)/12 $
  * $P_{A_{win}} = (3 - 2t + 2)/12 $
  * $P_{A_{win}} = (5 - 2t)/12 $
    * -- seems correct because it is 1/3 for t=0.5. 5/12 for t = 0 is at least a valid probability.
     
* If $t >= 1/2$
  * $P_{A_{win}} = (t/4) \times (1/3) + (t'/4) \times [1 + 2 - 2t' + (1-2t+4t^2)/3] $
  * $P_{A_{win}} = (t/4) \times (1/3) + (t'/4) \times [1 + 2t + (1 - 2t + 4t^2)/3] $
  * $P_{A_{win}} = (t/12) + ((1-t)/12) \times [3 + 6t + (1 - 2t + 4t^2)] $
  * $P_{A_{win}} = (t/12) + ((1-t)/12) \times [4 + 4t + 4t^2] $
  * $P_{A_{win}} = (t/12) + (1-t) (1 + t + t^2)/3 $
  * $P_{A_{win}} = (t/12) + (1-t^3)/3 $
  * $P_{A_{win}} = (4+t-4t^3)/12 $
    * -- seems correct because it is 1/3 for t=0.5. 1/12 for t = 1 is also a valid probability.

Now we need to integrate this over t.

$ \int_{0}^{1} P_{A_{win}} dt $

$ = 1/12 [ \int_{0}^{1/2}(5-2t) dt + \int_{1/2}^{1} (4+t-4t^3) dt ] $

$ = 1/12 [ (5t-t^2)\big|_0^{1/2} + (4t+t^2/2-t^4)\big|_{1/2}^1] $

$ = 1/12 [ ((5/2-1/4)-0) + ((7/2)-(2+1/8-1/16))] $

$ = 1/12 [ (9/4) + ((7/2)-(33/16))] $

$ = 1/12 [ (9/4) + ((56/16)-(33/16))] $

$ = 1/12 [ (36/16) + (23/16)] $

$ = 1/12 [ 59/16 ] $

$ = 59/192 $

---

Going to switch to a simple simulation approach, reusing code from last week.

In [3]:
def P_A_wins(a,b,c,t):
    a_sprints = a >= t
    b_sprints = b >= 0.5
    c_sprints = c >= 0.5
    if not a_sprints:
        if (b_sprints or c_sprints):
            return 0
        else :
            return 1/3
    else:
        if not b_sprints and not c_sprints:
            return 1
        if b_sprints and not c_sprints:
            return 1 if a > b else 0
        if c_sprints and not b_sprints: 
            return 1 if a > c else 0
        if b_sprints and c_sprints:
            if a > b and a > c:
                return 1
            else:
                return 0


# Trying the scipy integration - hopefully not too bad becuse this function has fewer discontinuities than last week's function.
from scipy import integrate
def my_integrate(eps=1e-2):
    integrate_4D = integrate.nquad(P_A_wins, [[0, 1], [0, 1], [0, 1], [0, 1]], opts={'epsabs': eps, 'epsrel': eps}, full_output=True)
    print(f"4D Integration Result with {eps=}: {integrate_4D}")

my_integrate(1e-1)
my_integrate(1e-2)
my_integrate(1e-3)
my_integrate(1e-4)
my_integrate(1e-5)

4D Integration Result with eps=0.1: (0.30572499879391596, 0.09375661992546208, {'neval': 7285383})
4D Integration Result with eps=0.01: (0.30725165589088593, 0.009988353222359622, {'neval': 23860389})
4D Integration Result with eps=0.001: (0.3072872671229215, 0.000999488213947873, {'neval': 79942653})
4D Integration Result with eps=0.0001: (0.3072907547973993, 9.996922276099551e-05, {'neval': 225246399})
4D Integration Result with eps=1e-05: (0.3072849867998079, 9.997327785615795e-06, {'neval': 14543303397})


Saved results:
<pre>4D Integration Result with eps=0.1: (0.30572499879391596, 0.09375661992546208, {'neval': 7285383})
4D Integration Result with eps=0.01: (0.30725165589088593, 0.009988353222359622, {'neval': 23860389})
4D Integration Result with eps=0.001: (0.3072872671229215, 0.000999488213947873, {'neval': 79942653})
4D Integration Result with eps=0.0001: (0.3072907547973993, 9.996922276099551e-05, {'neval': 225246399})
4D Integration Result with eps=1e-05: (0.3072849867998079, 9.997327785615795e-06, {'neval': 14543303397})</pre>

The first few runs (looser epsilon values) were pretty quick, but the runtime increased exponentially.

Conclusion
----

Oh my god!! I got the 2 approaches to line up. The analytical approach took about a hundred times more effort and a lot more time, but I guess the exactness is a whole different level.

**Fiddler answer = 5/8 = 0.625**

**Extra Credit answer = 59/192 = 0.30729166666666666666666666666667**